File: src/accounting/gst.py
# src/accounting/gst.py
"""
GST computation engine for SANSHIP ERP

Responsibilities:
- Determine GST type (CGST/SGST vs IGST)
- Aggregate GST amounts from invoice line items
- Produce ledger-ready tax breakup
- No DB writes
- No UI coupling
"""

from decimal import Decimal, ROUND_HALF_UP


# =====================================================
# HELPERS
# =====================================================
def _d(val) -> Decimal:
    try:
        return Decimal(str(val)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    except Exception:
        raise ValueError(f"Invalid numeric value: {val}")


# =====================================================
# GST CORE
# =====================================================
def compute_gst(
    items: list,
    supplier_state_code: str,
    customer_state_code: str
) -> dict:
    """
    Computes GST breakup for given invoice items.

    Args:
        items: list of invoice item dicts
        supplier_state_code: e.g. '27'
        customer_state_code: e.g. '27'

    Returns:
        {
            taxable_total,
            cgst: {rate, amount} | None,
            sgst: {rate, amount} | None,
            igst: {rate, amount} | None
        }
    """

    if not items:
        raise ValueError("No items provided for GST computation")

    same_state = supplier_state_code == customer_state_code

    taxable_total = Decimal("0.00")
    cgst_amt = Decimal("0.00")
    sgst_amt = Decimal("0.00")
    igst_amt = Decimal("0.00")

    cgst_rate = None
    sgst_rate = None
    igst_rate = None

    for it in items:
        taxable = _d(it.get("taxable_amount", 0))
        taxable_total += taxable

        if same_state:
            cgst_amt += _d(it.get("cgst_amt", 0))
            sgst_amt += _d(it.get("sgst_amt", 0))

            # capture rate once (audit reference)
            if cgst_rate is None:
                cgst_rate = _d(it.get("cgst_rate", 0))
            if sgst_rate is None:
                sgst_rate = _d(it.get("sgst_rate", 0))
        else:
            igst_amt += _d(it.get("igst_amt", 0))
            if igst_rate is None:
                igst_rate = _d(it.get("igst_rate", 0))

    result = {
        "taxable_total": float(taxable_total),
        "cgst": None,
        "sgst": None,
        "igst": None
    }

    if same_state:
        if cgst_amt > 0:
            result["cgst"] = {
                "rate": float(cgst_rate or 0),
                "amount": float(cgst_amt)
            }
        if sgst_amt > 0:
            result["sgst"] = {
                "rate": float(sgst_rate or 0),
                "amount": float(sgst_amt)
            }
    else:
        if igst_amt > 0:
            result["igst"] = {
                "rate": float(igst_rate or 0),
                "amount": float(igst_amt)
            }

    return result

File: SanshipMBL/src/main/kotlin/com/sanship/accounting/Gst.kt
package com.sanship.accounting

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * GST computation engine for SANSHIP ERP
 *
 * Responsibilities:
 * - Determine GST type (CGST/SGST vs IGST)
 * - Aggregate GST amounts from invoice line items
 * - Produce ledger-ready tax breakup
 * - No DB writes
 * - No UI coupling
 *
 * EXACT REPLICA of Python src/accounting/gst.py
 */
object Gst {
    
    // =====================================================
    // HELPERS
    // =====================================================
    // =====================================================
    // HELPERS
    // =====================================================
    
    val stateCodeMap = mapOf(
        "Jammu & Kashmir" to "01",
        "Himachal Pradesh" to "02",
        "Punjab" to "03",
        "Chandigarh" to "04",
        "Uttarakhand" to "05",
        "Haryana" to "06",
        "Delhi" to "07",
        "Rajasthan" to "08",
        "Uttar Pradesh" to "09",
        "Bihar" to "10",
        "Sikkim" to "11",
        "Arunachal Pradesh" to "12",
        "Nagaland" to "13",
        "Manipur" to "14",
        "Mizoram" to "15",
        "Tripura" to "16",
        "Meghalaya" to "17",
        "Assam" to "18",
        "West Bengal" to "19",
        "Jharkhand" to "20",
        "Odisha" to "21",
        "Chhattisgarh" to "22",
        "Madhya Pradesh" to "23",
        "Gujarat" to "24",
        "Dadra & Nagar Haveli and Daman & Diu" to "26",
        "Maharashtra" to "27",
        "Karnataka" to "29",
        "Goa" to "30",
        "Lakshadweep" to "31",
        "Kerala" to "32",
        "Tamil Nadu" to "33",
        "Puducherry" to "34",
        "Andaman & Nicobar Islands" to "35",
        "Telangana" to "36",
        "Andhra Pradesh" to "37",
        "Ladakh" to "38",
        "Other Territory" to "97",
        "Centre Jurisdiction" to "99"
    )
    
    fun getStateCode(stateName: String?): String? {
        if (stateName.isNullOrBlank()) return null
        // direct code check
        if (stateName.length == 2 && stateName.all { it.isDigit() }) return stateName
        // name lookup (case insensitive)
        return stateCodeMap.entries.find { it.key.equals(stateName, ignoreCase = true) }?.value
    }
    
    private fun toBigDecimal(value: Any?): BigDecimal {
        return try {
            BigDecimal(value.toString()).setScale(2, RoundingMode.HALF_UP)
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid numeric value: $value")
        }
    }
    
    // =====================================================
    // GST CORE
    // =====================================================
    data class GstResult(
        val taxableTotal: Double,
        val cgst: GstComponent? = null,
        val sgst: GstComponent? = null,
        val igst: GstComponent? = null
    )
    
    data class GstComponent(
        val rate: Double,
        val amount: Double
    )
    
    fun computeGst(
        items: List<Map<String, Any?>>,
        supplierStateCode: String?,
        customerStateCode: String?
    ): GstResult {
        /**
         * Computes GST breakup for given invoice items.
         *
         * Args:
         *     items: list of invoice item maps
         *     supplierStateCode: e.g. '27'
         *     customerStateCode: e.g. '27'
         *
         * Returns:
         *     GstResult with taxable_total and GST components
         */
        
        if (items.isEmpty()) {
            throw IllegalArgumentException("No items provided for GST computation")
        }
        
        val sameState = supplierStateCode == customerStateCode
        
        var taxableTotal = BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP)
        var cgstAmt = BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP)
        var sgstAmt = BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP)
        var igstAmt = BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP)
        
        var cgstRate: BigDecimal? = null
        var sgstRate: BigDecimal? = null
        var igstRate: BigDecimal? = null
        
        for (item in items) {
            val taxable = toBigDecimal(item["taxable_amount"] ?: 0)
            taxableTotal += taxable
            
            if (sameState) {
                cgstAmt += toBigDecimal(item["cgst_amt"] ?: 0)
                sgstAmt += toBigDecimal(item["sgst_amt"] ?: 0)
                
                // capture rate once (audit reference)
                if (cgstRate == null) {
                    cgstRate = toBigDecimal(item["cgst_rate"] ?: 0)
                }
                if (sgstRate == null) {
                    sgstRate = toBigDecimal(item["sgst_rate"] ?: 0)
                }
            } else {
                igstAmt += toBigDecimal(item["igst_amt"] ?: 0)
                if (igstRate == null) {
                    igstRate = toBigDecimal(item["igst_rate"] ?: 0)
                }
            }
        }
        
        return GstResult(
            taxableTotal = taxableTotal.toDouble(),
            cgst = if (sameState && cgstAmt > BigDecimal.ZERO) {
                GstComponent(
                    rate = (cgstRate ?: BigDecimal.ZERO).toDouble(),
                    amount = cgstAmt.toDouble()
                )
            } else null,
            sgst = if (sameState && sgstAmt > BigDecimal.ZERO) {
                GstComponent(
                    rate = (sgstRate ?: BigDecimal.ZERO).toDouble(),
                    amount = sgstAmt.toDouble()
                )
            } else null,
            igst = if (!sameState && igstAmt > BigDecimal.ZERO) {
                GstComponent(
                    rate = (igstRate ?: BigDecimal.ZERO).toDouble(),
                    amount = igstAmt.toDouble()
                )
            } else null
        )
    }
}
